[TRTLLM-6341][chore] Preliminary refactors on the kv cache manager before supporting swa kv cache reuse

eopXD opened the pull request:
## Description

<!--
Please explain the issue and the solution in short.
-->

This MR is a preliminary MR for implementing the SWA reuse mechanism for the kv cache manager. Please be aware that **no functional change is intended** in this merge request. The purpose of the clean-up is to decouple and remove existing functions for the up-coming SWA KV cache reuse change to be more natural and easier to review.

Right now, (1) streamLLM, and (2) beam search with SWA, are broken. We do not want to complicate the code base by stacking more features upon something that does not work. This MR prunes out the logic and add assertions so we can come back and re-support the broken feature and remove the assertion.

Since streamLLM (sink attention) is broken now, assertion is added under `KVCacheManager` ctor to guard for the value of `mSinkBlockTokenLength` and `mSinkBubbleLength`. Compute logics relate to it are pruned.

The beam search with SWA will still be broke when introducing the SWA KV cache reuse. We will revisit this problem in the future.

On top of this, we should make an effort to update the [supporting matrix](https://github.com/NVIDIA/TensorRT-LLM/blob/feat/1.0_doc_dev/docs/source/1.0/features/feature-combination-matrix.md) of the kv cache manager after merging the support of SWA KV cache reuse.

Changes are listed as following:
- Separate `KVCacheManager::updateToken` into `KVCacheManager::addToken` and `KVCacheManager::removeToken`. The functionality should be decoupled.
- Push utility `cacheSequenceBlockOffsets` and `cacheNewBlockOffset` from `KVCacheManager` down to `WindowBlockManager`. `KVCacheManager`-exposed functions should be real utilities that users of the structure can leverage. Implementation-detailed function calls should not exist at this level.
- Simplify "is shared last context block" logic under `KVCacheManager::addSequence`.

## Test Coverage

<!--
Please list clearly what are the relevant test(s) that can safeguard the changes in the PR. This helps us to ensure we have sufficient test coverage for the PR.
-->

Since no functional change is intended in this merge request, no test case is added. Several comments are added for future test coverage reminder.

For `LlmRequestTest.ParamTest`, `streaming=True` is commented out because we guard sink attention with assertion now.

In `capacitySchedulerTest`, `addToken` action to crossKVCacheManager is removed because in encoder-decoder model, generation tokens are added only to the decoder and not to the encoder.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Per-window KV-cache flows and a new explicit option to share the last context block; window-aware public APIs to update KV-cache offsets.

* **Refactor**
  * Centralized block allocation and offset-update responsibilities under window-aware block managers; per-window routing for sequence/token handling.

* **Behavior Changes**
  * Streaming generation disabled/guarded; legacy shared/unshared replacement paths removed; sequence/block APIs changed (window-aware, share-last flag).

* **Tests**
  * Unit tests simplified; sink-token >0 and multi-beam coverage deferred or disabled.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

eopXD committed:
[KV cache manager] No functional change intended, separate `KVCacheManager::updateToken` into `addToken` and `removeToken`

Given that streamLLM is broken now, removed related code logic and added
assertion to guard for sinkAttention settings. This allows us to revisit
the computation in the future.

Main reasons of this change:
Token addition and removal logic should be decoupled.
Flattens callstack and nested-if.
std::swap use is bizarre.
Broken features should be guarded safely.

Rename cache offset bookkeeping utility for clearance.
These utilities are pushed downwards from `KVCacheManager` to `BlockManager`.
- cacheBlockOffsets` --> `updateSequenceCacheBlockOffsets`
- cacheNewBlockOffsets` --> `updateLastCacheBlockOffsets`
- updateNewBlockPointer` --> `updateCacheBlockOffsetsAtIdx`

Comments have been added to test cases for reminder of future test
coverage on feature support.

Signed-off-by: eopXD <yuehtingc@nvidia.com>

eopXD committed:
[KV cache manager] No functional change intended, add comment for more context to distinguish `BlockManager::addSequence` and `WindowBlockManager::addSequence`

Signed-off-by: eopXD <yuehtingc@nvidia.com>

eopXD commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #14813](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/14813/) [ run ] triggered by Bot

tomeras91 reviewed (commented):
Overall looks good, although I do have a few comments/questions:
1. You say StreamingLLM (i.e. sink tokens) and beam search are not supported with SWA.. 
1.1. What broke this support?
1.2. You changed the unittests to reflect the fact they aren't supported. I assume this means these unittests passed before your changes (i.e. with sink tokens and beam search). So how did these unittests pass if it's not supported?

2. I asked in the PR comments as well.. is beam search really not supported at all? That's what it seems from the assertions you added, but you wrote in the description it's not supported only for SWA...


BTW - Great description! It was very helpful for reviewing

eopXD commented on cpp/tensorrt_llm/batch_manager/kvCacheManager.cpp:
```diff
@@ -1189,23 +1193,31 @@ void WindowBlockManager::addSequence(
+    {
+        allocateBlock(sequence, /*shareAmongBeams=*/true);
+    }
+    if (numContextBlocks > 1)
```
`numContextBlocks` will never be zero.

I am assuming that `inputLength` fed into `KVCacheManager::addSequence` will be greater than 0. `numContextBlocks` computed will be at least one block and at most the window size.

eopXD replied:
Code was updated to fix this as I was debugging for UT failures. Thank you for the catch.

eopXD commented on cpp/tensorrt_llm/batch_manager/kvCacheManager.cpp:
```diff
@@ -1868,93 +1830,33 @@ void KVCacheManager::cacheNewBlockOffsets(GenerationRequest& sequence, SizeType3
-        std::swap(currNumTokens, newNumTokens);
-    }
-
+    TLLM_CHECK_WITH_INFO(mSinkBlockTokenLength == 0 && mSinkBubbleLength == 0,
```
Good point. Code duplication here, its a note we revisit this component of the manager in the future to expand the functionality. Possibly, are you thinking of adding comment here for context instead of this assertion?

eopXD committed:
[KV cache manager] No functional change intended, simplify shared/unshared last context block logic under `KVCacheManager::addSequence`

The last context block is shared when:
- Operating on a cyclic kv cache and with no beam search
- Operating on cross kv cache
- Last context block is full

Prune incorrect comment left under `KVCacheManager::getNeededBlocksOneStep`.

No need to complicate things.

Signed-off-by: eopXD <yuehtingc@nvidia.com>

eopXD commented on cpp/tensorrt_llm/batch_manager/kvCacheManager.cpp:
```diff
@@ -2353,15 +2238,27 @@ BlocksPerWindow BaseKVCacheManager::calculateMaxNumBlocks(executor::KvCacheConfi
-    auto const beamWidth = sequence.getBeamWidth();
-
-    TLLM_CHECK_WITH_INFO(beamWidth == 1, "removeToken does not support beamWidth > 1");
+    if (sequence.getBeamWidth() > 1)
+    {
+        TLLM_LOG_WARNING("[KV cache manager] removeToken does not support beamWidth > 1");
+    }
```
Updated.

eopXD commented on cpp/tensorrt_llm/batch_manager/kvCacheManager.cpp:
```diff
@@ -2353,15 +2238,27 @@ BlocksPerWindow BaseKVCacheManager::calculateMaxNumBlocks(executor::KvCacheConfi
-    TLLM_CHECK_WITH_INFO(beamWidth == 1, "removeToken does not support beamWidth > 1");
+    if (sequence.getBeamWidth() > 1)
+    {
+        TLLM_LOG_WARNING("[KV cache manager] removeToken does not support beamWidth > 1");
```
Resolving this since warning is updated to assertion.

tomeras91 commented:
> Good question. What it seems to me now is that the KV cache manager manages all attentions with a "window size" specified. Since we are planning to support SWA KV cache reuse to grow linearly just like a full attention, the breakage for "beam search with SWA" also implies the breakage for "beam search with an infinite window, that is, full attention".
Beam search was added before SWA concept was introduced into KV cache manager, and I think except some UT-s we see here, we don't have other e2e coverage of it. So as it seems not ideal now, it is borderline-acceptable for me to prune out the feature and revisit it in the near future.

While it is true that we can think of full attention is SWA with infinite window, there is still an important difference. If the window size is equal to the maximum sequence length supported by the model, we're not actually in SWA. This is a fundamentally different scenario than actual SWA where the attention window is smaller than the maximum sequence length. 

Beam search is a pretty widely used feature and I would be very surprised if it is currently broken on `main`. Here are a few links to end2end tests using beam search. They should work as recent commits passed CI..
https://github.com/NVIDIA/TensorRT-LLM/blob/7e33ed6d61e30f8e1e4e2765d5d275726e7f0a6a/tests/unittest/_torch/test_beam_search.py
https://github.com/NVIDIA/TensorRT-LLM/blob/7e33ed6d61e30f8e1e4e2765d5d275726e7f0a6a/tests/integration/defs/accuracy/test_cli_flow.py#L79
https://github.com/NVIDIA/TensorRT-LLM/blob/7e33ed6d61e30f8e1e4e2765d5d275726e7f0a6a/tests/integration/defs/accuracy/test_llm_api_pytorch.py#L217

eopXD commented on cpp/tests/unit_tests/batch_manager/capacitySchedulerTest.cpp:
```diff
@@ -453,7 +453,8 @@ TEST_F(CapacitySchedulerTest, SimpleShouldFit)
     auto capacitySchedulerPolicies = std::vector<CapacitySchedulerPolicy>{CapacitySchedulerPolicy::kGUARANTEED_NO_EVICT,
         CapacitySchedulerPolicy::kMAX_UTILIZATION, CapacitySchedulerPolicy::kSTATIC_BATCH};
-    auto sinkTokenLens = std::vector<SizeType32>{0, 4};
+    // TODO: Support and add coverage for sinkTokenLen > 0. (e.g. 4)
+    auto sinkTokenLens = std::vector<SizeType32>{0};
     for (auto capacitySchedulerPolicy : capacitySchedulerPolicies)
```
No, prefer current labeling.

eopXD commented on cpp/tensorrt_llm/batch_manager/kvCacheManager.cpp:
```diff
@@ -1868,93 +1830,33 @@ void KVCacheManager::cacheNewBlockOffsets(GenerationRequest& sequence, SizeType3
+    TLLM_CHECK_WITH_INFO(mSinkBlockTokenLength == 0 && mSinkBubbleLength == 0,
+        "[kv cache manager] streamLLM is not supported at the moment");
+    auto& sequence = getSequence(requestId);
+    TLLM_CHECK_WITH_INFO(sequence.getBeamWidth() == 1, "[kv cache manager] addToken does not support beamWidth > 1");
```
Beam search assertion is removed now to keep current {full attnetion+beam search} behavior workable.

eopXD commented:
/bot run --disable-fail-fast

eopXD commented:
@tomeras91 Looking at the code closely, I don't think the combination of {beam search + full attention + no reuse} works now.

[Looking at current code](https://github.com/NVIDIA/TensorRT-LLM/blob/main/cpp/tensorrt_llm/batch_manager/kvCacheManager.cpp#L1925) under `KVCacheManager::addToken`,
when we need a new block, the a block is created as long as `newNumTokens <= maxTokenNum`.
For full attention, we will always be creating a new block and not cycling back.
Notice that its an `else-if` that is following the block creation to replace the shared block with unshared.
We will need to replace shared block with unshared blocks when we are doing beam search in full attention.

This is incorrect code. Does the tests you have mentioned above being actively ran, if so, I am spectical about the test coverage.

eopXD commented on cpp/tensorrt_llm/batch_manager/kvCacheManager.cpp:
```diff
@@ -2353,15 +2309,23 @@ BlocksPerWindow BaseKVCacheManager::calculateMaxNumBlocks(executor::KvCacheConfi
 {
+    // TODO: add streamLLM support
     auto& sequence = getSequence(requestId);
-    auto const beamWidth = sequence.getBeamWidth();
-
-    TLLM_CHECK_WITH_INFO(beamWidth == 1, "removeToken does not support beamWidth > 1");
     if (sequence.getNumTokens() == 0)
     {
         return;
     }
-    updateToken(sequence, false);
+    TLLM_CHECK_WITH_INFO(sequence.getBeamWidth() == 1, "[kv cache manager] removeToken does not support beamWidth > 1");
+    sequence.removeTokens(1);
+    for (auto const [windowSize, metadata] : mBlockManager.getWindowSizesMetadata())
+    {
+        SizeType32 const maxTokensInWindow = metadata.maxTokenNum;
+        SizeType32 const tokensInWindow = sequence.getNumTokens() % maxTokensInWindow;
+        if (tokensInWindow % getTokensPerBlock() == 0 && tokensInWindow <= maxTokensInWindow)
+        {
+            mBlockManager.releaseLastBlock(sequence, windowSize);
+        }
+    }
 }
```
I agree with you. I hate dead codes too, will submit MR to remove the ussage.

thorjohnsen reviewed (commented):
Mostly looks good to me, my only concern is whether the changes to addSequence will work with SWVA.

tomeras91 reviewed (approved):
Overall LGTM. Added a small comment about redundant if statement

tensorrt-cicd commented:
[PR_Github #15094](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15094/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #11396](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/11396/) completed with status: 'FAILURE'

eopXD commented on cpp/tensorrt_llm/batch_manager/kvCacheManager.cpp:
```diff
@@ -1989,25 +1961,7 @@ void KVCacheManager::addSequence(
-
-        // Consider the temporaryAttentionWindow when allocating blocks.
-        auto const effectiveInputLength = std::min(inputLength, maxTokenNum + temporaryAttentionWindow);
+        auto const effectiveInputLength = std::min(inputLength, windowSize);
```
I am more than happy to know more context of temporaryAttentionWindow. Its a code creeping all over the codebase with not much context or explanation.

thorjohnsen replied:
I tried to get an explanation for temporaryAttentionWindow, but nobody seems to know anything about it. Most code contributed towards temporaryAttentionWindow seems to have been part of a restructuring effort were the goal was to effectuate no functional change for temporaryAttentionWindow. I don't know what the right answer is, but I'm pretty sure it's not what is in your MR. It should work for full attention layers and sliding window is kind of work-in-progress so maybe we can ignore this for now.

eopXD replied:
This does not sound right to me.

Going to recover the code for now. It is worthy to make an effort to remove this in the future. It is bad to keep unclear variable(semantic) in the code and trying to stack new features on it.

eopXD commented on cpp/tensorrt_llm/batch_manager/kvCacheManager.cpp:
```diff
@@ -1189,23 +1193,31 @@ void WindowBlockManager::addSequence(
+    {
+        allocateBlock(sequence, /*shareAmongBeams=*/true);
+    }
+    if (numContextBlocks >= 1)
```
Yes indeed. Updated.

eopXD commented:
/bot run --disable-fail-fast

eopXD commented on cpp/tensorrt_llm/batch_manager/kvCacheManager.cpp:
```diff
@@ -1885,76 +1895,37 @@ void KVCacheManager::updateNewBlockPointer(GenerationRequest& sequence, SizeType
                 mBlockManager.replaceSharedBlock(sequence, windowSize, nextBlockIdx);
-                updateNewBlockPointer(sequence, windowSize, nextBlockIdx);
+                mBlockManager.updateCacheBlockOffsetsAtIdx(sequence, windowSize, nextBlockIdx);
             }
```
I was mistaken. The newly allocated blocks under addToken is default as unshared. The if-condition below is code that does not work since SWA+beam search is not good now.

eopXD commented:
/bot run --disable-fail-fast

eopXD committed:
[batch manager][unit test] No functional change intended, remove kv cache block allocation on token addition for cross kv manager

For an encode-decoder model, generation tokens are added only to the
decoder and not the encoder.

Signed-off-by: eopXD <yuehtingc@nvidia.com>

tensorrt-cicd commented:
[PR_Github #15413](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15413/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #15412](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15412/) [ run ] completed with state `ABORTED`

tensorrt-cicd commented:
[PR_Github #15697](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15697/) [ run ] triggered by Bot

eopXD commented on cpp/tensorrt_llm/batch_manager/kvCacheManager.cpp:
```diff
@@ -1885,76 +1895,37 @@ void KVCacheManager::updateNewBlockPointer(GenerationRequest& sequence, SizeType
-                if (addToken)
-                {
-                    mBlockManager.allocateBlock(sequence, windowSize);
-                    cacheNewBlockOffsets(sequence, windowSize);
-                }
-                else
-                {
-                    mBlockManager.releaseLastBlock(sequence, windowSize);
-                }
+                // Allocate new unshared blocks until the window can always
+                // accommodate "window size" number of tokens.
+                mBlockManager.allocateBlock(sequence, windowSize);
+                mBlockManager.updateLastCacheBlockOffsets(sequence, windowSize);
             }
             else if (sequence.getBeamWidth() > 1)
             {
-                TLLM_CHECK_WITH_INFO(addToken, "Remove token is not supported with beam search");
-                // Get next block index
-                SizeType32 nextBlockIdx = nextTokenIdxInCache / getTokensPerBlock();
-                // Replace the shared block with the unshared ones
+                // For beam search, shared block is replaced with unshared ones
+                auto const nextBlockIdx = (sequence.getNumTokens() - 1) / getTokensPerBlock();
                 mBlockManager.replaceSharedBlock(sequence, windowSize, nextBlockIdx);
-                updateNewBlockPointer(sequence, windowSize, nextBlockIdx);
+                mBlockManager.updateCacheBlockOffsetsAtIdx(sequence, windowSize, nextBlockIdx);
             }
         }
     }
 }
```
Good point and nice observation. This part of code will be changed further in the SWA kv cache reuse support. So for now, not doing change to this.
